[Excel] (Range) Add formula precedents article section (#2253)
OfficeDev/office-js-docs-pr: docs/excel/excel-add-ins-ranges-advanced.md

You can also find the cell responsible for spilling into a given cell by using the [Range.getSpillParent](/javascript/api/excel/excel.range#getspillparent--) method. Note that `getSpillParent` only works when the range object is a single cell. Calling `getSpillParent` on a range with multiple cells will result in an error being thrown (or a null range being returned for `Range.getSpillParentOrNullObject`).

> ## Trace formula precedents

Reviewer: Does "Trace" refer to the UI behavior of drawing the lines between formula cells and their precedents? If so, maybe we can rename this section to "Get formula precedents".

(Note that "trace" is used a couple times below, so if this gets changed, please change the other uses too.)

Author: Trace is the term on the UI button ("Trace precedents"), which on click draws arrows from precedent cells to the currently selected cell. I think it's important to identify in this section that the API is related Trace precedents button + behavior. Do you agree? I see your point about changing the title of the section, because this API doesn't draw arrows.

Reviewer: While "trace" does likely refer to the drawing of arrows, I agree with your point that we should tie the API to the UI button. Do you think it would be worth noting in the text that our APIs don't draw any arrows?

Author: Yes, good point!

> An Excel formula often refers to other cells or groups of cells. These cells or groups of cells are known as "precedents" to the formula. To learn more about Excel features related to relationships between cells, see the [Display the relationships between formulas and cells](https://support.microsoft.com/office/display-the-relationships-between-formulas-and-cells-a59bef2b-3701-46bf-8ff1-d3518771d507) article. 

Author: The SOC article avoids this grammatical construction (precedents to/of a formula) entirely, and my other searches indicate that *only* Microsoft uses the terms formula and precedent together. The closest aligning phrase in the SOC article is "cells that provide data to a formula (precedents)" which is how I ended up using 'to' instead of 'of'. Is there anywhere else you suggest I search? My inclination is to make sure I'm consistent but that this is a gray area, or language in development. Perhaps it's a grammatical construction to avoid entirely, but I think dancing around the language makes the concept more confusing.

> An Excel formula often refers to other cells or groups of cells. These cells or groups of cells are known as "precedents" to the formula. To learn more about Excel features related to relationships between cells, see the [Display the relationships between formulas and cells](https://support.microsoft.com/office/display-the-relationships-between-formulas-and-cells-a59bef2b-3701-46bf-8ff1-d3518771d507) article. 

Reviewer: The place you hear "precedent" the most is in legal documentation. You would say "a legal precedent". Would it make sense to frame it similarly here? Like, "These cells or groups of cells are known as formula "precedents"." ?

> An Excel formula often refers to other cells or groups of cells. These cells or groups of cells are known as "precedents" to the formula. To learn more about Excel features related to relationships between cells, see the [Display the relationships between formulas and cells](https://support.microsoft.com/office/display-the-relationships-between-formulas-and-cells-a59bef2b-3701-46bf-8ff1-d3518771d507) article. 

Author: Yes, I agree! I will rephrase this section and others to avoid the awkward preposition.

> With [Range.getDirectPrecedents](/javascript/api/excel/excel.range#getdirectprecedents--), your add-in can locate the direct precedents to a formula. The `getDirectPrecedents` method can trace precedents to a formula within a single worksheet and across worksheets within a single workbook. This method can't trace precedent cells across workbooks. 

Reviewer: Saying "single worksheet" and "single workbook" makes it sound like there's a choice developers will make when calling this API. Instead, they'll always get everything in the workbook. Consider condensing this section.
```suggestion
With [Range.getDirectPrecedents](/javascript/api/excel/excel.range#getdirectprecedents--), your add-in can locate the direct precedents to a formula. The `getDirectPrecedents` method returns the precedents of a formula that are in the same workbook. This method can't trace precedent cells across workbooks. 
```

> With [Range.getDirectPrecedents](/javascript/api/excel/excel.range#getdirectprecedents--), your add-in can locate the direct precedents to a formula. The `getDirectPrecedents` method can trace precedents to a formula within a single worksheet and across worksheets within a single workbook. This method can't trace precedent cells across workbooks. 

Reviewer: I really like the callout that this won't work across workbooks. Consider promoting that information to a NOTE or IMPORTANT.

Author: Since the UI feature *can* trace cells across workbooks, I'll switch it to IMPORTANT.

> With [Range.getDirectPrecedents](/javascript/api/excel/excel.range#getdirectprecedents--), your add-in can locate the direct precedents to a formula. The `getDirectPrecedents` method can trace precedents to a formula within a single worksheet and across worksheets within a single workbook. This method can't trace precedent cells across workbooks. 

Reviewer: Can we describe the returned `WorkbookRangeAreas` object here? It's a complex data structure that could use some object model context. Maybe something like "`Range.getDirectPrecedents` returns a `WorkbookRangeAreas` object. This object contains the address of all the precedents in the workbook. It has a separate `RangeAreas` object for each worksheet containing at least one precedent for the formula. See "Work with multiple ranges simultaneously in Excel add-ins" for more information on working with the `RangeAreas` object."

Saying all that, it might also be worth adding a section about WorkbookRangeAreas to the multiple ranges article. I'll leave that to your discretion.

Author: Thanks for the explanation! I will look into adding a new section to the multiple ranges article.

> The following sample calls the `getDirectPrecedents` method on the active range, loads the addresses of the direct precedent cells, and then loops through each precedent cell and changes its background color to yellow. 

Reviewer: Let's focus on the behavior instead of the code in this preamble. It's always focus make the scenario sound more like the problem they're trying to solve or the UI behavior.
```suggestion
The following sample gets the direct precedents for the active range and then changes the background color of those precedent cells to yellow. 
```

> [!NOTE]

> > The active range must contain a formula that references other cells in the same workbook for this sample to work properly. 

Reviewer: Does this sample crash if it doesn't have a cell with precedents selected? If not, I'd say not highlighting anything is working properly :-)

Author: The sample doesn't crash, it just doesn't highlight anything. I'll adjust the note.

```js
Excel.run(function (context) {
    // Precedents are cells referenced by the formula in a cell.
    var range = context.workbook.getActiveCell();
    var directPrecedents = range.getDirectPrecedents();
    range.load("address");
    directPrecedents.areas.load("address");

>     return context.sync();

Reviewer: This `return` ends the function. You'll need to chain the second half of the function together with a `then()`. The dynamic arrays sample on this page demonstrates that pattern.

    console.log(`Direct precedent cells of ${range.address}:`);

    // Use the direct precedents API to loop through precedents of the active cell.
    for (var i = 0; i < directPrecedents.areas.items.length; i++) {
      // Highlight and print out the address of each precedent cell.
      directPrecedents.areas.items[i].format.fill.color = "Yellow";
      console.log(`  ${directPrecedents.areas.items[i].address}`);
    }
    return context.sync();
}).catch(errorHandlerFunction);
```

## See also

- [Work with ranges using the Excel JavaScript API](excel-add-ins-ranges.md)
- [Excel JavaScript object model in Office Add-ins](excel-add-ins-core-concepts.md)
- [Work with multiple ranges simultaneously in Excel add-ins](excel-add-ins-multiple-ranges.md)

> - [Display the relationships between formulas and cells](https://support.microsoft.com/office/display-the-relationships-between-formulas-and-cells-a59bef2b-3701-46bf-8ff1-d3518771d507)

Reviewer: We don't have the other SOC topics linked down here. I think the logic behind that it that if the want to learn about the specific feature correlated with an API, they can check out SOC, but the average developer trying to navigate our doc set won't want to.

Author: I see your point. The other Support docs referenced in this article are not linked in this See also section. I added the link here because a Support doc is linked in the Comments article See also section (I was working on both in tandem and cross-referenced for style and structure). If the general rule is that we don't reference Support docs in the See also section, should I remove the Support reference from the Comments article See also section?

Reviewer: I think it's okay in the Comments article because it's targeted to a single feature. "Range" is a broad collection of Excel features, so links to SOC (support.office.com) wouldn't be clearly applicable.
